src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs

```diff
@@ -22,329 +21,185 @@ namespace System.Threading
 
-            if (millisecondsTimeout >= -1)
+            if (result == ObjectHeader.HeaderLockResult.Success)
             {
-                EnterHelperResult tryEnterResult = TryEnter_FastPath_WithTimeout(obj, millisecondsTimeout);
-                if (tryEnterResult == EnterHelperResult.Entered)
-                {
-                    return true;
-                }
-                else if (millisecondsTimeout == 0 && (tryEnterResult == EnterHelperResult.Contention))
-                {
-                    return false;
-                }
+                return;
             }
 
-            return TryEnter_Slowpath(obj, millisecondsTimeout);
-        }
-
-        // The JIT should inline this method to allow check of lockTaken argument to be optimized out
-        // in the typical case. Note that the method has to be transparent for inlining to be allowed by the VM.
-        public static void TryEnter(object obj, int millisecondsTimeout, ref bool lockTaken)
-        {
-            if (lockTaken)
-                ThrowLockTakenException();
-
-            ArgumentNullException.ThrowIfNull(obj, null);
+            if (result == ObjectHeader.HeaderLockResult.Failure)
+            {
+                throw new SynchronizationLockException();
+            }
 
-            TryEnter_Timeout_WithLockTaken(obj, millisecondsTimeout, ref lockTaken);
+            GetLockObject(obj).Exit();
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
         public static bool IsEntered(object obj)
         {
-            ArgumentNullException.ThrowIfNull(obj);
+            ObjectHeader.HeaderLockResult result = ObjectHeader.IsAcquired(obj);
+            if (result == ObjectHeader.HeaderLockResult.Success)
+                return true;
+
+            if (result == ObjectHeader.HeaderLockResult.Failure)
+                return false;
 
-            return IsEnteredNative(obj);
+            return GetLockObject(obj).IsHeldByCurrentThread;
         }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern bool IsEnteredNative(object obj);
-
-        /*========================================================================
-    ** Waits for notification from the object (via a Pulse/PulseAll).
-    ** timeout indicates how long to wait before the method returns.
-    ** This method acquires the monitor waithandle for the object
-    ** If this thread holds the monitor lock for the object, it releases it.
-    ** On exit from the method, it obtains the monitor lock back.
-    **
-        ** Exceptions: ArgumentNullException if object is null.
-    ========================================================================*/
-        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Monitor_Wait")]
-        [return: MarshalAs(UnmanagedType.Bool)]
-        private static partial bool Wait(ObjectHandleOnStack obj, int millisecondsTimeout);
+        #endregion
+
+        #region Public Wait/Pulse methods
```

Comment:
Everything from here still the end of the file looks identical to NAOT. Move to shared?